Correlate traces to profiles with the OTel ebpf profiler

elsakeirouz opened the pull request:
# Description
Based on Elastic's work to correlate traces to profiles for Java using the ebpf profiler.

elsakeirouz commented:
What do we think about using something like `#ifdef linux/unix` for parts that might not be relevant on Windows/MacOS? To be fair, since this code specifically is meant to work with the ebpf profiler on Linux, and will eventually only run if prompted, I do not see the point of compiling it on other OSs.

dubloom commented:
> What do we think of using something like `#ifdef linux/unix` for parts that might not be relevant on Windows/MacOS? To be fair, since this code specifically is meant to work with the ebpf profiler on Linux, and will eventually only run if prompted, I do not see the point of compiling it on other OSs.

My 2 cents here: if the code does not add a big compile time, I think we will pay more in code readibility by adding #ifdef (as the code is split into multiple files) than we will win in compile time. It can be worth adding some comments to tell it will only work in Linux though.
(Note this is my personal opinion and I'm not really a contributor of dd-trace-cpp atm).

elsakeirouz commented:
> > What do we think of using something like `#ifdef linux/unix` for parts that might not be relevant on Windows/MacOS? To be fair, since this code specifically is meant to work with the ebpf profiler on Linux, and will eventually only run if prompted, I do not see the point of compiling it on other OSs.
> 
> My 2 cents here: if the code does not add a big compile time, I think we will pay more in code readibility by adding #ifdef (as the code is split into multiple files) than we will win in compile time. It can be worth adding some comments to tell it will only work in Linux though. (Note this is my personal opinion and I'm not really a contributor of dd-trace-cpp atm).

Noted, thanks for your input!

Commit:
feat: expose process storage for trace to profile correlation

* include/datadog/tracer.h
* include/datadog/tracer_signature.h
* src/datadog/tracer.cpp

Commit:
feat: add data in tls for trace to profile correlation

* BUILD.bazel
* cmake/compiler/clang.cmake
* cmake/compiler/gcc.cmake
* include/datadog/tls_storage.h
* include/datadog/trace_segment.h
* include/datadog/tracer.h
* src/datadog/span.cpp
* src/datadog/trace_segment.cpp
* src/datadog/tracer.cpp

elsakeirouz commented:
FYI, the global variables' naming is not set in stone. We plan on moving towards something more generic that does not use the elastic name.

Additionally, I am in the process of a writing a problem statement regarding the way we share process data since dd-trace-cpp already shares that data through memfds.

Commit:
feat: add option to enable correlation

* include/datadog/environment.h
* include/datadog/tracer.h
* include/datadog/tracer_config.h
* src/datadog/tracer.cpp
* src/datadog/tracer_config.cpp

Commit:
feat: add tests for trace-to-profile correlation

* test/test_tracer.cpp
* test/test_tracer_config.cpp

elsakeirouz commented on test/test_tracer.cpp:
```diff
@@ -1566,6 +1566,32 @@ TEST_CASE("128-bit trace IDs") {
+    auto finalized_config = finalize_config(config);
+    REQUIRE(finalized_config);
+    Tracer tracer{*finalized_config};
+    REQUIRE(elastic_apm_profiling_correlation_process_storage_v1 != nullptr);
+    auto span = tracer.create_span();
+    REQUIRE(elastic_apm_profiling_correlation_tls_v1 != nullptr);
```
Could add more checks here for the actual stored values.

elsakeirouz commented:
What's left to do here is figure how we want to handle runtime-ids in case of forks.

dmehala committed:
workaround

Commit:
fix: init storage before telemetry init

The telemetry initialization now takes a few seconds.
This delay is problematic for trace-to-profile correlation
since it delays the initialization of the process storage,
which the eBPF profiler checks once when it first encounters
the process.

* src/datadog/tracer.cpp

Commit:
feat: add some documentation

* README.md

elsakeirouz commented on test/test_tracer.cpp:
```diff
@@ -1566,6 +1566,48 @@ TEST_CASE("128-bit trace IDs") {
+    auto finalized_config = finalize_config(config);
+    REQUIRE(finalized_config);
+    Tracer tracer{*finalized_config};
+    REQUIRE(elastic_apm_profiling_correlation_process_storage_v1 != nullptr);
```
Since we have not yet defined what the process storage is going to look like, I am not validating the data that is stored in there. Once that is specified, we can look into decoding the contents.

r1viollet commented on cmake/compiler/clang.cmake:
```diff
@@ -5,6 +5,26 @@
+  else()
+    message(FATAL_ERROR "Only aarch64 and x86-64 are supported (found: ${CMAKE_SYSTEM_PROCESSOR})")
+  endif()
+
```
This part could be conditioned to an expected $TLS_DIALECT set above

ivoanjo commented on src/datadog/span.cpp:
```diff
@@ -40,6 +41,16 @@ Span::~Span() {
+  // When a span is finished, we must update the span_id to its parent's.
+  if (elastic_apm_profiling_correlation_process_storage_v1 != nullptr &&
+      parent_id().has_value()) {
+    elastic_apm_profiling_correlation_tls_v1->valid = 0;
+    elastic_apm_profiling_correlation_tls_v1->span_id = parent_id().value();
+    elastic_apm_profiling_correlation_tls_v1->valid = 1;
```
For future reference, if anyone's curious, I was reviewing this with @elsakeirouz and we noticed we're missing some synchronization here, [per the spec](https://github.com/elastic/apm/blob/bd5fa9c1/specs/agents/universal-profiling-integration.md#concurrency-safe-updates):

> Note that APM-agents must make sure that compilers do not reorder the steps listed above.
